File: common/src/main/java/io/github/notenoughupdates/moulconfig/common/IFontRenderer.java
package io.github.notenoughupdates.moulconfig.common;

import io.github.notenoughupdates.moulconfig.common.text.StructuredText;
import org.jspecify.annotations.NullMarked;

import java.util.List;

@NullMarked
public interface IFontRenderer {
    int getHeight();

    default int getStringWidth(String string) {
        return getStringWidth(StructuredText.of(string));
    }

    int getStringWidth(StructuredText structuredText);

    default int getCharWidth(char c) {
        return getStringWidth(Character.toString(c));
    }

    List<StructuredText> splitText(StructuredText structuredText, int width);

    default List<StructuredText> splitLines(StructuredText structuredText) {
        return splitText(structuredText, Integer.MAX_VALUE);
    }

    default String trimStringToWidth(String string, int width) {
        return trimStringToWidth(string, width, false);
    }

    String trimStringToWidth(String string, int width, boolean reversed);
}

File: common/src/main/java/io/github/notenoughupdates/moulconfig/common/IFontRenderer.kt
package io.github.notenoughupdates.moulconfig.common

import io.github.notenoughupdates.moulconfig.common.text.StructuredText

/**
 * Not for manual implementation. This should be implemented by the corresponding platform.
 */
interface IFontRenderer {
    val height: Int
    fun getStringWidth(string: StructuredText): Int
    fun getStringWidth(string: String): Int = getStringWidth(StructuredText.of(string))
    fun getCharWidth(char: Char): Int
    fun splitText(text: StructuredText, width: Int): List<StructuredText>
    fun splitLines(text: StructuredText): List<StructuredText> {
        return splitText(text, Int.MAX_VALUE)
    }

    fun trimStringToWidth(string: String, width: Int) = trimStringToWidth(string, width, false)
    fun trimStringToWidth(string: String, width: Int, reverse: Boolean): String

}
